src/login.php

```diff
@@ -25,40 +25,202 @@
+                $cookieInfo[ 'path' ],
+                $cookieInfo[ 'domain' ],
+                $cookieInfo[ 'secure' ],
+                true
+            );
+        } else {
+            setcookie($cookieName, $cookieValue, $cookieExpire);
+        }
+    }
+
+    private static function checkResult(\mysqli_stmt $stmt)
+    {
+        global $userID, $loggedin, $_language;
+    
+        if ($stmt->bind_result($userID, $language, $lastlogin)
+                && $stmt->fetch()) {
+            $loggedin = true;
+            $_SESSION['ws_user'] = $userID;
+            $_SESSION['ws_lastlogin'] = $lastlogin;
+
+            if (!empty($language) && isset($_language)) {
+                if ($_language->setLanguage($language)) {
+                    $_SESSION['language'] = $language;
+                }
+            }
+        }
+    }
+
+    /**
+     * Set a cookie that can be used for a persistent login. The login cookie
+     * is set to expire after a certain interval.
+     * The cookie has the form <userID>:<key>, with <key> being a random
+     * 64-character string (base64-encoded). Apart from setting the cookie, an
+     * entry containing the user ID, a cryptographic hash of the key and the
+     * expiration time will be inserted in the database.
+     */
+    public static function set($cookieName, $user, $expiration)
+    {
+        global $_database;
+
+        $key  = self::generateKey();
+        $hash = self::generateHash($key);
+        $cookieValue = $user . ":" . base64_encode($key);
+        $cookieExpire = $expiration > 0 ? time() + $expiration : 0;
+
+        $stmt = $_database->prepare("INSERT INTO " . PREFIX . "cookies
+                (userID, cookie, expiration) VALUES (?, ?, ?)");
+        if (false !== $stmt) {
+            $success = $stmt->bind_param("isi", $user, $hash, $cookieExpire)
+                && $stmt->execute();
+            $stmt->close();
+        }
+
+        if ($success) {
+            self::setCookie($cookieName, $cookieValue, $cookieExpire);
+        }
+    }
+
+    /**
+     * Remove a login cookie. Both the cookie on the client side and the
+     * database entry on the server side are removed. Note that this does not
+     * remove the client session.
+     */
+    public static function clear($cookieName)
+    {
+        global $_database;
+
+        $cookie = $_COOKIE[$cookieName];
 
-if (isset($_SESSION['ws_auth'])) {
-    if (stristr($_SESSION['ws_auth'], "userid") === false) {
-        $authent = explode(":", $_SESSION['ws_auth']);
+        if (isset($cookie)) {
+            $authent = explode(":", $cookie);
+            $user = $authent[0];
+            $key  = base64_decode($authent[1]);
 
+            $stmt = $_database->prepare("DELETE FROM " . PREFIX . "cookies
+                    WHERE `userID` = ? AND `cookie` = ?");
+            if (false !== $stmt) {
+                $stmt->bind_param('is', $user, self::generateHash($key)) &&
+                $stmt->execute();
```

Comment:
I think it is not needed to use && header.
to increase readability it would be nice if you split the lines